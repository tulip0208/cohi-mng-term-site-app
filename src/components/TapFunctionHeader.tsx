/**-------------------------------------------
 * 共通_機能・アプリ種別ヘッダ・タップ機能
 * components/TapFunctionHeader.tsx
 * ---------------------------------------------*/
import React, {useState, useEffect} from 'react';
import {TouchableOpacity, Text, View} from 'react-native';
import {styles} from '../styles/CommonStyle'; // 共通スタイル
import {logScreen, logUserAction} from '../utils/Log';
import {StackNavigationProp} from '@react-navigation/stack';
import {RootList} from '../navigation/AppNavigator'; // RootList のインポート
import {useNavigation} from '@react-navigation/native';
type NavigationProp = StackNavigationProp<RootList, 'WA1050'>;
interface Props {
  appType: string;
  viewTitle: string;
  functionTitle: string;
  sourceScreenId: string;
}

const TapFunctionHeader = ({
  appType,
  viewTitle,
  functionTitle,
  sourceScreenId,
}: Props) => {
  const [tapCount, setTapCount] = useState<number>(0);
  const [timerId, setTimerId] = useState<NodeJS.Timeout | null>(null);
  const navigation = useNavigation<NavigationProp>();
  useEffect(() => {
    if (tapCount === 5) {
      // 画面遷移後はタップカウンターをリセット
      setTapCount(0);
      if (timerId) {
        clearTimeout(timerId);
      }
      logUserAction('ボタン押下:端末設定');
      logScreen('画面遷移: WA1050_端末設定');
      navigation.navigate('WA1050', {sourceScreenId: sourceScreenId}); // 画面遷移のロジック
    }
  }, [tapCount, timerId, navigation, sourceScreenId]);

  const handleTitleTap = () => {
    console.log('count tap ', tapCount + 1);
    // タイマーがすでにある場合はリセット
    if (timerId) {
      clearTimeout(timerId);
    }

    // タップカウンターを更新
    setTapCount(prevCount => prevCount + 1);

    // 5秒後にタップカウンターをリセットするタイマーを設定
    const newTimerId = setTimeout(() => {
      if (tapCount < 5) {
        setTapCount(0);
        console.log('count tap reset 0');
      }
    }, 5000);

    setTimerId(newTimerId);
  };

  return (
    <TouchableOpacity onPress={handleTitleTap}>
      <View style={styles.functionHeaderContainer}>
        <Text style={styles.functionHeaderLeft}>{appType}</Text>
        <Text style={styles.functionHeaderMiddle}>{viewTitle}</Text>
        <Text style={styles.functionHeaderRight}>{functionTitle}</Text>
      </View>
    </TouchableOpacity>
  );
};

export default TapFunctionHeader;
