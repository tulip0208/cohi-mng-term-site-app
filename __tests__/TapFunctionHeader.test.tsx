import React from 'react';
import { TouchableOpacity } from 'react-native';
import TestRenderer, { act, ReactTestRenderer } from 'react-test-renderer';
import { NavigationContext } from '@react-navigation/native';
import TapFunctionHeader from '../src/components/TapFunctionHeader.tsx';

// ログ関数とナビゲーションのモック
jest.mock('../src/utils/Log.tsx', () => ({
  logScreen: jest.fn(),
  logUserAction: jest.fn()
}));

const mockNavigate = jest.fn();

// テスト用のナビゲーションコンテキスト
const navigationContext: any = {
  navigate: mockNavigate,
};

describe('TapFunctionHeader コンポーネント', () => {
  // 初期化処理
  jest.useFakeTimers();
  let consoleSpy: jest.SpyInstance, clearTimeoutSpy: jest.SpyInstance, testRenderer: ReactTestRenderer;

  // テスト前に実行される関数
  beforeEach(() => {
    // console.log のモック化
    consoleSpy = jest.spyOn(console, 'log').mockImplementation();
    clearTimeoutSpy = jest.spyOn(global, 'clearTimeout');

    // testRendererの初期化
    testRenderer = TestRenderer.create(
      <NavigationContext.Provider value={navigationContext}>
        <TapFunctionHeader appType="App" viewTitle="View" functionTitle="Function" sourceScreenId="Source" />
      </NavigationContext.Provider>
    );
  });

  // テスト後に実行される関数
  afterEach(() => {
    // モックのクリーンアップ
    consoleSpy.mockRestore();
    clearTimeoutSpy.mockRestore();
  });

  ///////----------------------------------------///////
  ///////----ここ以下に必要なテストを記述する-------///////
  ///////----------------------------------------///////

  it('正しくレンダリングされること', () => {
    const tree = testRenderer.toJSON();
    // 正しいかをスナップショットで比較をする
    expect(tree).toMatchSnapshot();
  });

  it('タップがカウントされること', () => {
    // タップしてカウントアップ
    act(() => {
      const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
      touchableOpacity.props.onPress();
    });
    // 1のカウントが行われていること
    expect(consoleSpy).toHaveBeenCalledWith("count tap ", 1);
    act(() => {
      const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
      touchableOpacity.props.onPress();
    });
    // 2のカウントが行われていること
    expect(consoleSpy).toHaveBeenCalledWith("count tap ", 2);
  });

  it('タップをカウントし、5秒後にリセットされること', () => {
    // タップしてカウントアップ
    act(() => {
      const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
      touchableOpacity.props.onPress();
    });
    // 1のカウントが行われていること
    expect(consoleSpy).toHaveBeenCalledWith("count tap ", 1);


    // 5秒後の動作確認
    act(() => {
      jest.advanceTimersByTime(5000);
    });
    // カウントリセットのログが流れること
    expect(consoleSpy).toHaveBeenCalledWith("count tap reset 0");
    // タップしてカウントアップ
    act(() => {
      const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
      touchableOpacity.props.onPress();
    });
    // リセットされ、1のカウントが行われていること
    expect(consoleSpy).toHaveBeenCalledWith("count tap ", 1);
  });

  it('タップをカウントし、4秒後にリセットされていないこと', () => {
    // タップしてカウントアップ
    act(() => {
      const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
      touchableOpacity.props.onPress();
    });
    // 1のカウントが行われていること
    expect(consoleSpy).toHaveBeenCalledWith("count tap ", 1);


    // 4秒後の動作確認
    act(() => {
      jest.advanceTimersByTime(4000);
    });
    // カウントリセットのログが流れていないこと
    expect(consoleSpy).not.toHaveBeenCalledWith("count tap reset 0");
    // タップしてカウントアップ
    act(() => {
      const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
      touchableOpacity.props.onPress();
    });
    // 2のカウントが行われていること
    expect(consoleSpy).toHaveBeenCalledWith("count tap ", 2);
  });

  it('タイマーがない場合はリセットが行われないこと', () => {
    const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
    // 1回目タップ
    act(() => {
      touchableOpacity.props.onPress();
    });
    // タイマーが存在しないためリセットが行われないこと
    expect(clearTimeoutSpy).not.toHaveBeenCalled();
  });

  it('すでにタイマーがある時はリセットを行うこと', () => {
    const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
    // 1回目タップ
    act(() => {
      touchableOpacity.props.onPress();
    });
    // 1回目はタイマーが存在しないためリセットが行われないこと
    expect(clearTimeoutSpy).not.toHaveBeenCalled();
    act(() => {
      jest.advanceTimersByTime(1000);
    });
    // 2回目タップ
    act(() => {
      touchableOpacity.props.onPress();
    });
    // 2回目はタイマーが存在するためリセットが行われること
    expect(clearTimeoutSpy).toHaveBeenCalled();

  });

  it('4回タップしてもナビゲートが起こらないこと', () => {
    // 4回タップ
    const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
    act(() => {
      for (let i = 0; i < 4; i++) {
        touchableOpacity.props.onPress();
      }
    });
    // タイマーのリセットが行われていないか
    expect(clearTimeoutSpy).not.toHaveBeenCalled();
    // ナビゲーションが呼ばれていないか
    expect(mockNavigate).not.toHaveBeenCalledWith('WA1050', { sourceScreenId: 'Source' });
  });

  it('5回タップするとナビゲートが起こるか', () => {
    // 5回タップ
    act(() => {
      const touchableOpacity = testRenderer.root.findByType(TouchableOpacity);
      for (let i = 0; i < 5; i++) {
        touchableOpacity.props.onPress();
      }
    });
    // タイマーのリセットが行われているか
    expect(clearTimeoutSpy).toHaveBeenCalled();
    // ナビゲーションが呼ばれているか
    expect(mockNavigate).toHaveBeenCalledWith('WA1050', { sourceScreenId: 'Source' });
  });
});
