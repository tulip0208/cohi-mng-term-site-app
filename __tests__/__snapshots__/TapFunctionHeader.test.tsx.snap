// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TapFunctionHeader コンポーネント 正しくレンダリングされること 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1E74BD",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": 58.59375,
        "justifyContent": "space-between",
        "padding": 4.6875,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "flex": 1,
          "fontFamily": "ipaexg",
          "fontSize": 30.46875,
          "fontWeight": "bold",
          "justifyContent": "center",
          "marginLeft": 11.71875,
          "textAlign": "left",
        }
      }
    >
      App
    </Text>
    <Text
      style={
        {
          "alignSelf": "center",
          "color": "white",
          "flex": 1,
          "fontFamily": "ipaexg",
          "fontSize": 35.15625,
          "fontWeight": "bold",
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
    >
      View
    </Text>
    <Text
      style={
        {
          "color": "white",
          "flex": 1,
          "fontFamily": "ipaexg",
          "fontSize": 30.46875,
          "fontWeight": "bold",
          "justifyContent": "center",
          "marginRight": 11.71875,
          "textAlign": "right",
        }
      }
    >
      Function
    </Text>
  </View>
</View>
`;
